### DISCUSSION TITLE
Best way to instantiate a Web3Provider and how to access.

### DISCUSSION THREAD

**Discussion by michaeljohnbennett** [wevm/wagmi#272] [Q&A]
I am working on integrating the WAGMI components into a project that uses the Web3-react components but also relies heavily on the Web3Provider from ethers.

I have found it hard to get a solid/simple example of using this within the wagmi components and have come up with the following solution but i wanted to know if it was the best way?

# 1. Define the Provider in App.tsx

```typescript
import Dashboard from "./Dashboard";

import {Provider, chain, defaultChains, Connector} from 'wagmi';
import {InjectedConnector} from 'wagmi/connectors/injected';
import {WalletConnectConnector} from 'wagmi/connectors/walletConnect';

import {getDefaultProvider, providers as ethproviders} from "ethers";
import { getNetwork } from "@ethersproject/providers";

const defaultChain = chain.mainnet;

const getProvider = (_config: { chainId?: number; connector?: Connector }) => {
  let wProvider: any = window.ethereum;
  console.log("getProvider", {wProvider, winEth: window.ethereum});
  let ret: any;
  if(!wProvider){
    ret = getDefaultProvider(getNetwork(_config.chainId ?? defaultChain.id));
  } else {
    ret = new ethproviders.Web3Provider(wProvider);
  }
  console.log("getProvider.returning", {wProvider, winEth: window.ethereum, ret});
  return ret;
};

const connectors = [
  new InjectedConnector({chains: defaultChains}),
  new WalletConnectConnector({
    chains: defaultChains,
    options: {
      qrcode: true,
    },
  }),
];

function App() {
  return (
    <Provider connectors={connectors} provider={getProvider}>
      <Dashboard/>
    </Provider>
  );
}

export default App;

```

So this tries to get the `window.ethereum` provider and if it doesn't find one it tries to use the default provider instead.

Now that seems to work ok but I then wanted to access the provider later on as a Web3Provider but the hooks for `useProvider` only return the `BaseProvider` type. I had to resort to using the `useConnect` hook instead and then I can access the Web3Provider to call `send(...)` for RPC calls.

# 2. Access the Web3Provider

So now I have to access the provider like so:

```typescript
  const [{data: connectData}] = useConnect();
  const provider = connectData.connector?.getProvider();

 // ... a bit later on
 const sendAsync = promisify(provider.sendAsync.bind(provider));
  try {
    return await sendAsync(payload);
  } catch (error) {
    console.error(error);
    return {
      jsonrpc: payload.jsonrpc,
      id: payload.id,
      error
    };
  }
```

If I had used `useProvider` here it wouldn't work. Why would that be? It kind of looks like it should.

I realise this isn't probably the normal usecase most people use this for but all the bits are kinda there, it just seems a bit more typing black magic and you can possibly get a better/strongly typed provider. Web3React components allow typings on retrieval of the provider etc from the hooks. This made things a lot easier to work with but we wanted to use WAGMI to get the walletconnect code in as well.

Any pointers would be appreciated.

**Comment by tmm** [comment]
Working on adding better type inference for the ethers provider.

**Comment by searchableguy** [comment]
+1 for this. Thanks for the solution. I ran into the same issue while passing provider.
